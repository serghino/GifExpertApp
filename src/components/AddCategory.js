import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {

    // this state should contain a value by default.
    const [inputValue, setInputValue] = useState(''); // ''

    const handleInputChange = ( e ) => {
        setInputValue( e.target.value );
    }

    const handleSubmit = (e) => {
        // prevent auto refresh of the main page.
        e.preventDefault();

        if ( inputValue.trim().length > 2 ) {
            setCategories( cats => [ inputValue, ...cats, ] );
            setInputValue('');
        }

    }

    return (
        <form onSubmit={ handleSubmit }>
            <input 
                type="text"
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>
    )
}


AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
