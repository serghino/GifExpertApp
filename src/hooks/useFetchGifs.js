import { useState, useEffect } from 'react'
import { getGifs } from '../helpers/getGifs';


export const useFetchGifs = ( category ) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true,
        total: 0
    });
    // Similar to componentDidMount and componentDidUpdate:
    useEffect( () => {
        // this must be sync.
        getGifs( category )
            .then( ({gifs,total}) => {
                setState({
                    data: gifs,
                    loading: false,
                    total
                });
            })

    }, [category])
    return state; // { data:[], loading: true };
}


