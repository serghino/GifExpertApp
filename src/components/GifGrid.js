import React from 'react';
import PropTypes from 'prop-types'
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    const { data:images, loading, total } = useFetchGifs( category );

    return (
        <>
            <div className="gridTitle">
                <h3 className="animate__animated animate__fadeIn"> { category } </h3>
                <p><b>Total</b>: {total}</p>
            </div>

            { loading && <p className="animate__animated animate__flash">Loading</p> }

            <div className="card-grid">
                
                {
                    images.map( img => (
                        <GifGridItem 
                            key={ img.id }
                            { ...img }
                        />
                    ))
                }
            
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
}