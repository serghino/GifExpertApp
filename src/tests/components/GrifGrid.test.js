import React from "react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";
// mock useFetchGifs.
jest.mock("../../hooks/useFetchGifs");

describe("Test component <GifGrid />", () => {
  const category = "simpsons";

  test("should show <GifGrid /> correctly", () => {
   useFetchGifs.mockReturnValue({
    data: [],
    loading: true,
    total: 0
   });
   const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("should show up the items from category value using Fetch Gifs", () => {
    const gifs = [{
        id: '123',
        url: 'http://localhost',
        title: 'ABC'
    }];
    
    useFetchGifs.mockReturnValue({
        data: gifs,
        loading: false,
        total: 1
       });
       const wrapper = shallow(<GifGrid category={category} />);
       expect(wrapper).toMatchSnapshot();
       expect(wrapper.find('.animate__animated animate__flash').exists()).toBe(false);
       expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
  });   
});
