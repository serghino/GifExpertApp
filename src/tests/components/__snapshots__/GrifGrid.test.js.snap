// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component <GifGrid /> should show <GifGrid /> correctly 1`] = `
<Fragment>
  <div
    className="gridTitle"
  >
    <h3
      className="animate__animated animate__fadeIn"
    >
       
      simpsons
       
    </h3>
    <p>
      <b>
        Total
      </b>
      : 
      0
    </p>
  </div>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Test component <GifGrid /> should show up the items from category value using Fetch Gifs 1`] = `
<Fragment>
  <div
    className="gridTitle"
  >
    <h3
      className="animate__animated animate__fadeIn"
    >
       
      simpsons
       
    </h3>
    <p>
      <b>
        Total
      </b>
      : 
      1
    </p>
  </div>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="123"
      key="123"
      title="ABC"
      url="http://localhost"
    />
  </div>
</Fragment>
`;
